scripts/babysitter

```diff
@@ -0,0 +1,364 @@
+	formatter_class=argparse.RawTextHelpFormatter)
+argparser.add_argument('-s', '--signal', dest='signal', metavar='signal', default='TERM',
+	help="Send this signal to the command on timeout, instead of TERM.")
+argparser.add_argument('-k', '--kill-after-duration', dest='kill_after', metavar='duration',
+	help="If process continues running after signal, send KILL after this duration.")
+argparser.add_argument('duration',
+	help="Time to run before sending signal.")
+argparser.add_argument('command', nargs="+", help="Command+args to run.")
+
+# Repeatedly parse the given args until we find the shortest prefix for which parsing succeeds.
+argc = len(sys.argv)
+extra_args = []
+for limit in range(1,argc+1):
+	try:
+		if limit == argc: # On the final pass, parse for real
+			argparser.hesitating = False
+		args = argparser.parse_args(sys.argv[1:limit])
+		# If we're still here, parse_args succeeded.
+		# The final parsed arg is the command; remaining argv items are its args.
+		extra_args = sys.argv[limit:]
+		break
+	except Hesitate: # Before the final pass, if parse_args fails, skip
+		pass
+
+argparser.hesitating = False # Just act like a normal argparser from here
+
+durationre = re.compile(r'(\d+)([smhd]?)')
+def parse_duration(duration): # Accept units
+	match = durationre.match(duration)
+	if not match:
+		argparser.error("Could not understand duration %s" % duration)
+	time, units = match.group(1), match.group(2)
+	time = int(time)
+	if units == 'm':
+		time *= 60
+	elif units == 'h':
+		time *= 60*60
+	elif units == 'd':
+		time *= 60*60*24
+	return time
+
+def parse_signal(sig):        # Accept names
+	if sig.isdigit():
+		return int(sig)
+	for k,v in signal.__dict__.iteritems():
+		if k == ("SIG%s" % sig):
+			return v
+	argparser.error("Could not understand signal name %s" % sig)
+
+# Final interpretation of arguments
+duration = parse_duration(args.duration)
+kill_after = parse_duration(args.kill_after) if args.kill_after is not None else None
+timeout_signal = parse_signal(args.signal)
+command = args.command + extra_args
+
+# Process environment
+global_env = copy.deepcopy( os.environ )
+
+verbose = VERBOSE_KEY in global_env
+logging = LOGGING_DIR_KEY in global_env
+logging_dir = global_env[LOGGING_DIR_KEY] if logging else None
+logfile = os.path.join(logging_dir, LOGGING_FILE) if logging else None
+crash_resuming = CRASH_RESUME_KEY in global_env
+failmax = int(global_env[RETRY_KEY]) if RETRY_KEY in global_env else 0
+babysitting = logging or failmax
+if babysitting:
+	babysitter_dir = tempfile.mkdtemp()
+	global_env[CURRENT_TEST_KEY] = os.path.join(babysitter_dir, CURRENT_TEST_FILE)
+	global_env[RAN_TEST_KEY]     = os.path.join(babysitter_dir, RAN_TEST_FILE)
+	global_env[FAILED_TEST_KEY]  = os.path.join(babysitter_dir, FAILED_TEST_FILE)
+
+### Utility functions
+
+def wait(proc, duration):
+	# FIXME: If we can guarantee Python 3.3, Popen objects have a wait(timeout) method
```

Comment:
I think I was thinking that if ANY of our test machines at some point get python 3.3, we could add a `version_info` check and get improved (more precise) behavior on those machines.